src/placement/mapless_jump_map.c

```diff
@@ -0,0 +1,990 @@
+
+/**
+ * This struct is used to to hold information while
+ * finding rebuild targets for shards located on unavailable
+ * targets.
+ */
+struct remap_node {
+	uint32_t rank;
+	uint32_t shard_idx;
+	uint32_t fseq;
+};
+
+
+/**
+ * Helper function used for sorting rebuild list by fail sequence
+ * This function compares the fail sequence of the remap_nodes
+ */
+static int
+map_rebuild_cmp(void *array, int a, int b)
+{
+	struct remap_node *sort = (struct remap_node *)array;
+
+	if (sort[a].fseq > sort[b].fseq)
+		return 1;
+	if (sort[a].fseq  < sort[b].fseq)
+		return -1;
+	return 0;
+}
+
+/**
+ * Helper function used for sorting rebuild list by fail sequence
+ * This function swaps the two remap_nodes.
+ */
+static void
+map_rebuild_swap(void *array, int a, int b)
+{
+	struct remap_node *sort;
+	struct remap_node  tmp;
+
+	sort = ((struct remap_node *)array);
+
+	tmp = sort[a];
+	sort[a] = sort[b];
+	sort[b] = tmp;
+}
+
+/** Function pointers for sorting the rebuild list. */
+static daos_sort_ops_t map_rebuild_sops = {
+	.so_cmp		= map_rebuild_cmp,
+	.so_swap	= map_rebuild_swap,
+};
+
+/**
+ * This function sets a specific bit in the bitmap
+ * It expects the bitmap to be zero indexed from left
+ * to right.
+ */
+static inline void
+set_bit(uint8_t *bitmap, uint64_t bit)
+{
+        uint64_t offset = bit / 8;
+        uint8_t position = bit % 8;
+        bitmap[offset] |= (0x80 >> position);
+}
+
+/** Returns the bit at a specific position in the bitmap */
+static inline uint8_t
+get_bit(uint8_t *bitmap, uint64_t bit)
+{
+	uint64_t offset = bit / 8;
+        uint8_t position = bit % 8;
+        return ((bitmap[offset] & (0x80 >> position)) != 0);
+}
+
+/**
+ * Returns true or false depending on whether all bits given in a range
+ * are set or not, the range is inclusive. This is used to in the bookkeeping
+ * to determine if all children in a domain have been used.
+ *
+ * \return	returns 0 if the entire range does not contain set bits and
+ *		returns 1 if all bits are set
+ */
+static inline uint8_t
+is_range_set(uint8_t *bitmap, uint64_t start, uint64_t end)
+{
+    uint8_t mask = 0xFF >> (start % 8);
+
+    if (end >> 3 == start >> 3) {
+        mask &= (0xFF << (8-((end%8)+1)));
+        return (bitmap[(start >> 3)] & mask) == mask;
+    }
+
+    if ((bitmap[(start >> 3)] & mask) != mask)
+        return 0;
+
+    mask = (0xFF << (8-((end%8)+1)));
+    if((bitmap[(end >> 3)] & mask) != mask)
+        return 0;
+
+    if ((end >> 3) > ((start >> 3) + 1)) {
+        int i;
+        for(i = (start >> 3) + 1; i < (end >> 3); ++i)
+            if(bitmap[i] != 0xFF)
+                return 0;
+    }
+
+    return 1;
+}
+
+/**
+ * This function clears a continuous range of bits in the bitmap.
+ * The range if from start to end inclusive.
+ */
+static inline void
+clear_bitmap_range(uint8_t *bitmap, uint64_t start, uint64_t end)
+{
+        uint8_t mask = (0xFF >> (start % 8));
+
+        mask = ~mask;
+        if (end >> 3 == start >> 3) {
```

Comment:
(style)  code indent should use tabs where possible
(style)  suspect code indent for conditional statements (8, 12)